### PR TITLE
Fix for solveCobraNLP (matlab), testFitC13Data, and test for solveCobraNLP

### PR DESCRIPTION
**Pull request by laurentheirendt**
**Major changes**:

- Fix for `matlab` interface in `solveCobraNLP`
- Fix for issue #129 
- Eventual fix for issue #333 
- Added test for `FitC13Data`
- Added test for `solveCobraNLP`

**I hereby confirm that I have:**

- [X] Tested my code on my own machine
- [X] Followed the guidelines in the [contributing](https://github.com/opencobra/cobratoolbox/blob/master/.github/CONTRIBUTING.md) document
- [X] Followed the [code styleguide](https://github.com/opencobra/cobratoolbox/blob/master/.github/CONTRIBUTING.md#code-styleguide)
- [X] Checked to ensure there aren't other open [Pull Requests](https://github.com/opencobra/cobratoolbox/pulls) for the same update/change
- [X] Written platform-independent code
- [X] Used `pwd` to get the current directory
- [X] Used `filesep` for paths (e.g., `['myPath' filesep 'myFile.m']`)
- [X] Made sure that computational results are reproducible
- [X] Written a sensible test according to the [Test styleguide](https://github.com/opencobra/cobratoolbox/blob/master/.github/CONTRIBUTING.md#test-styleguide)
- [X] Tested your PR locally prior to submission
- [X] Checked that all tests pass
- [X] Tested the code on Linux

### CODE REVIEW

**Comment by laurentheirendt** [comment]
Agree, I will fix this, although we cannot test it just yet.

**Comment by laurentheirendt** [comment]
@tpfau, may you move `test/testOptimizeCbModelNLP/SimpleQPObjective.m` and `test/testOptimizeCbModelNLP/testOptimizeCbModelNLP.m` to `/test/verifiedTests/testSolvers` ?

**Comment by tpfau** [comment]
Moved the tests

File: test/verifiedTests/testSolvers/testOptimizeCbModelNLP.m [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+load('ecoli_core_model','model')
+
+% define the solver packages to be used to run this test
+solverPkgs = {'glpk'};  % tomlab_snopt
+
+for k = 1:length(solverPkgs)
+
+    % change the COBRA solver (NLP)
+    solverOK = changeCobraSolver(solverPkgs{k});
```
Comment (syarra) [inline test/verifiedTests/testSolvers/testOptimizeCbModelNLP.m]: I would be 'matlab' instead (to use `fmincon`).

File: test/verifiedTests/testSolvers/SimpleQPObjective.m [hunk 2]
```diff
@@ -0,0 +1,9 @@
+function [ function_value] = SimpleQPObjective( values, Problem)
+%SIMPLELPOBJECTIVE Optimizes a simple QP objective for an input.
```
Comment (syarra) [inline test/verifiedTests/testSolvers/SimpleQPObjective.m]: `SIMPLEQPOBJECTIVE`

File: test/verifiedTests/testSolvers/NLPscripts/NLP_objFunction.m [hunk 3]
```diff
@@ -0,0 +1,11 @@
+function f = NLP_objFunction(x, Prob)
+% f = NLP_objFunction(x, Prob)
+% From tomlab quickguide
```
Comment (syarra) [inline test/verifiedTests/testSolvers/NLPscripts/NLP_objFunction.m]: This problem is called the Rosenbrock problem. We should add this in the doctring.

File: src/solvers/solveCobraNLP.m [hunk 4]
```diff
@@ -1,310 +1,365 @@
...
+if ~isfield(NLPproblem, 'b')
+    NLPproblem.b = zeros(size(A, 1), 1);
+end
+
+if isfield(NLPproblem,'csense')
+    [b, csense] = deal(NLPproblem.b, NLPproblem.csense);
+elseif isfield(NLPproblem,'b_U')
+    [b_L,b_U] = deal(NLPproblem.b_L,NLPproblem.b_U);
+else
+    display('either .b_U, .b_L or .b, .csense must be defined')
```
Comment (syarra) [inline src/solvers/solveCobraNLP.m]: Should this throw an error?

File: src/solvers/NLPobjPerFlux.m [hunk 5]
```diff
@@ -11,15 +11,13 @@
 % Prob          NLP problem structure
 %
 %OUTPUT
-% value         Objective flux / v.^2
+% value         Objective flux / sum(v.^2)^
```
Comment (syarra) [inline src/solvers/NLPobjPerFlux.m]: Missing `(1/4)`

File: src/solvers/NLPobjPerFlux.m [hunk 6]
```diff
@@ -11,15 +11,13 @@
...
 % c wasn't defined as written so added Prob as input to define c from the
 % model by Daniel Zielinski 3/19/10
 
-model = Prob.user.model;
+c = Prob.objArguments{1};
 
-c = model.c == 1;
-
-value = -sum(c.*fluxVector)/sum(fluxVector.^2);
+value = sum(c.*fluxVector)/sum(fluxVector.^2);
```
Comment (syarra) [inline src/solvers/NLPobjPerFlux.m]: This function does not compute what is in the docstring!

File: test/verifiedTests/testSolvers/testOptimizeCbModelNLP.m (hunk 1 above)
Comment (tpfau) [inline test/verifiedTests/testSolvers/testOptimizeCbModelNLP.m]: The default functions use glpk to determine a basic solution for initialization of the NLP optimization.

File: src/solvers/solveCobraNLP.m (hunk 4 above)
Comment (tpfau) [inline src/solvers/solveCobraNLP.m]: probably.

File: src/solvers/NLPobjPerFlux.m (hunk 5 above)
Comment (tpfau) [inline src/solvers/NLPobjPerFlux.m]: Argh, that was me.... need to remove it again... sry

File: src/solvers/NLPobjPerFlux.m (hunk 6 above)
Comment (tpfau) [inline src/solvers/NLPobjPerFlux.m]: again, it's me who is to blame.

Comment (tpfau) [inline src/solvers/NLPobjPerFlux.m]: Ok.... 
solveCobraNLP expects a minimisation (according to the documentation). Simultaneously optimizeCbModelNLP allows to set whether its a minimsation or maximisation.
Question: should we modify solveCobraNLP to allow both min and maximisations (Wouldn't be so difficult), or should we remove the option from optimzeCbModelNLP? The problem is, that we have function names which are being called (which are expected to be provided as strings, which makes a prior +/- addition impossible.

File: src/solvers/NLPobjPerFlux.m (hunk 5 above)
Comment (tpfau) [inline src/solvers/NLPobjPerFlux.m]: Ok, I updated the PR adding osense as a required field for the NLPproblem input of solveCobraNLP. That addresses definitions of objective functions but does not address the "warning" output, which should probably be an error.

**Comment by tpfau** [comment]
Ok, I updated the PR adding osense as a required field for the NLPproblem input of solveCobraNLP. That addresses definitions of objective functions but does not address the "warning" output, which should probably be an error.

File: test/verifiedTests/testSolvers/testOptimizeCbModelNLP.m (hunk 1 above)
Comment (syarra) [inline test/verifiedTests/testSolvers/testOptimizeCbModelNLP.m]: Where is this done? I cannot find it!

Comment (tpfau) [inline test/verifiedTests/testSolvers/testOptimizeCbModelNLP.m]: src/randomObjFBASol <-- This function calls optimizeCbModel to get a solution and is used as default initialisation function for optimizeCbModelNLP.

**Pull request merged**